### PR TITLE
[ETFE-3123] /movements filters

### PR DESCRIPTION
**Pull request by james-work-account**

### CODE REVIEW

File: app/controllers/ViewAllMovementsController.scala [hunk 1]
```diff
@@ -16,74 +16,110 @@
...
-            form = formProvider().fill(searchOptions),
-            action = routes.ViewAllMovementsController.onSubmit(ern, searchOptions),
-            ern = ern,
-            movements = movementList.movements,
-            sortSelectItems = MovementSortingSelectOption.constructSelectItems(Some(searchOptions.sortBy.code)),
-            searchSelectItems = MovementSearchSelectOption.constructSelectItems(searchOptions.searchKey.map(_.code)),
-            pagination = paginationHelper.constructPagination(pageCount, ern, searchOptions)
-          ))
-        }
+      if (searchOptions.index <= 0 || searchOptions.index > pageCount) {
```
Comment (machinesleet) [inline app/controllers/ViewAllMovementsController.scala:89]: Is it worth a small comment explaining why this guard is here, and the reason for the redirect?

File: app/controllers/ViewAllMovementsController.scala [hunk 2]
```diff
@@ -16,74 +16,110 @@
...
+      movementList <- EitherT(getMovementListConnector.getMovementList(ern, Some(searchOptions)).map {
+        case Left(NotFoundError) => Right(GetMovementListResponse(Seq(), 0))
+        case value => value
+      })
+
+      epcs <- EitherT(getExciseProductCodesConnector.getExciseProductCodes())
+      exciseProductCodeOptions = MovementListSearchOptions.CHOOSE_PRODUCT_CODE +: epcs
+
+      countries <- EitherT(getMemberStatesConnector.getMemberStates())
+      sortedCountries = MovementListSearchOptions.CHOOSE_COUNTRY +: countries.sortBy(_.displayName)
```
Comment (machinesleet) [inline app/controllers/ViewAllMovementsController.scala:84]: Perhaps `selectCountryOptions` would be clearer here?

File: app/models/MovementFilterStatusOption.scala [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    case ManuallyClosed.code => ManuallyClosed
+    case PartiallyRefused.code => PartiallyRefused
+    case Refused.code => Refused
+    case Replaced.code => Replaced
+    case Rejected.code => Rejected
+    case Stopped.code => Stopped
+    case invalid => throw new IllegalArgumentException(s"Invalid argument of '$invalid' received which can not be mapped to a MovementFilterStatusOption")
+  }
+
+  def selectItems(existingAnswer: Option[MovementFilterStatusOption])(implicit messages: Messages): Seq[SelectItem] =
```
Comment (machinesleet) [inline app/models/MovementFilterStatusOption.scala:104]: Looks like this method isn't covered in `MovementFilterStatusOptionSpec`

File: app/models/MovementFilterUndischargedOption.scala [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+object MovementFilterUndischargedOption extends Enumerable.Implicits {
+
+  case object Undischarged extends WithName("Y") with MovementFilterUndischargedOption {
+    override val displayName = "viewAllMovements.filters.undischarged.undischarged"
+  }
+
+  val values: Seq[MovementFilterUndischargedOption] = Seq(Undischarged)
+
+  def checkboxItems(implicit messages: Messages): Seq[CheckboxItem] =
+    values.zipWithIndex.map {
```
Comment (machinesleet) [inline app/models/MovementFilterUndischargedOption.scala:39]: bit picky, but do you need `zipWithIndex` for one value?

File: app/models/MovementListSearchOptions.scala [hunk 5]
```diff
@@ -47,41 +109,173 @@ case class MovementListSearchOptions(searchKey: Option[MovementSearchSelectOptio
...
     new QueryStringBindable[MovementListSearchOptions] {
 
-      override def bind(key: String, params: Map[String, Seq[String]]): Option[Either[String, MovementListSearchOptions]] =
+      override def bind(key: String, params: Map[String, Seq[String]]): Option[Either[String, MovementListSearchOptions]] = {
         Some(for {
           sortOrder <- stringBinder.bind("sortBy", params).getOrElse(Right(ArcAscending.code))
           index <- intBinder.bind("index", params).getOrElse(Right(DEFAULT_INDEX))
           searchKey <- stringBinder.bind("searchKey", params).map(_.map(Some(_))).getOrElse(Right(None))
           searchValue <- stringBinder.bind("searchValue", params).map(_.map(Some(_))).getOrElse(Right(None))
+          traderRole <- stringBinder.bind("traderRole", params).map(_.map(Some(_))).getOrElse(Right(None))
```
Comment (machinesleet) [inline app/models/MovementListSearchOptions.scala:142]: Could you write a small helper method for the calls on each of these lines after `stringBinder.bind`?

File: app/models/MovementListSearchOptions.scala [hunk 6]
```diff
@@ -47,41 +109,173 @@ case class MovementListSearchOptions(searchKey: Option[MovementSearchSelectOptio
...
+             dateOfReceiptFrom: Option[LocalDate],
+             dateOfReceiptTo: Option[LocalDate]
+           ): MovementListSearchOptions = {
+
+    val traderRole: Option[MovementFilterDirectionOption] = {
+      (traderRoleOptions.contains(MovementFilterDirectionOption.GoodsIn), traderRoleOptions.contains(MovementFilterDirectionOption.GoodsOut)) match {
+        case (true, true) => Some(MovementFilterDirectionOption.All)
+        case (true, _) => Some(MovementFilterDirectionOption.GoodsIn)
+        case (_, true) => Some(MovementFilterDirectionOption.GoodsOut)
+        case _ => None
```
Comment (machinesleet) [inline app/models/MovementListSearchOptions.scala:210]: Looks like these cases aren't covered in `MovementListSearchOptionsSpec`

File: app/models/MovementListSearchOptions.scala [hunk 7]
```diff
@@ -47,41 +109,173 @@ case class MovementListSearchOptions(searchKey: Option[MovementSearchSelectOptio
...
+      }
+    }
+
+    val undischargedMovements = undischargedMovementsOptions.toSeq match {
+      case Seq(MovementFilterUndischargedOption.Undischarged) => Some(MovementFilterUndischargedOption.Undischarged)
+      case _ => None
```
Comment (machinesleet) [inline app/models/MovementListSearchOptions.scala:216]: Looks like this case isn't covered. 
Also, is there a reason for the `.toSeq` call on `undischargedMovementsOptions`?

File: test/controllers/ViewAllMovementsControllerSpec.scala
Comment (machinesleet) [inline test/controllers/ViewAllMovementsControllerSpec.scala:1]: This file has become a bit of a beast. Could the get/post blocks be their own files? Or some helper methods for the views (which are largely the same)

Comment (james-work-account) [reply test/controllers/ViewAllMovementsControllerSpec.scala:1]: Happy to extract the views into helper methods, I agree this file is a little large now...

File: test/viewmodels/helpers/MovementPaginationHelperSpec.scala [hunk 8]
```diff
@@ -31,12 +31,12 @@ class MovementPaginationHelperSpec extends SpecBase with MovementListFixtures {
   object Helper extends MovementPaginationHelper()
 
   def createPageItem(index: Int): PaginationItem = PaginationItem(
-    href = onPageLoad(testErn, MovementListSearchOptions(Some(ARC), Some("ARC123456"), ArcAscending, index, 10)).url,
+    href = onPageLoad(testErn, MovementListSearchOptions(Some(ARC), Some("ARC123456"), ArcAscending, None, None, None, None, None, None, None, None, None, index, 10)).url,
```
Comment (machinesleet) [inline test/viewmodels/helpers/MovementPaginationHelperSpec.scala:34]: Probably wotth having a default `MovementListSearchOptions` instance at the top of the test, and just changing the `index` and `maxRows` fields where relevant

**Review by machinesleet — APPROVED** [review]

**Pull request merged**